### PR TITLE
Support sending f1 to typed f4 addresses

### PR DESCRIPTION
**Pull request by cypt4**
Resolves https://github.com/brave/brave-browser/issues/30401
Support sending f1->f4.
1) FilAddress is extended to support new f4 namespace. f4 address consists of 2 identifiers - agent id and address.
We currently support only agent id "10".
2) fvm_shared lib is updated to 3.3.0 to support new f4 addresses to sign.
2) When f4 address is selected as recipient "Method" field should be changed  to 3844450837 for EstimateGas and MPoolPush operations.

Update:
fvm_shared lib is forked to brave-org repo (brave-experiments until PR is approved).
Library was stripped to depend on old versions: https://github.com/brave-experiments/ref-fvm/pull/2

Update:
Upstream of the fvm_shared lib is used

Sec review: 
https://github.com/brave/security/issues/1290

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
1) Create Filecoin account, obtain some funds
2) Take default Ethereum account and translate it address to f410 address using https://beryx.zondax.ch/address_converter
3) Use obtained f410 address when send from Filecoin account
4) Check that transaction succeeds, blockexplorer correctly shows info, account has the transaction in the history

### CODE REVIEW

File: build/rust/Cargo.lock
Comment (fmarier) [inline build/rust/Cargo.lock:209]: cc @kdenhartog for new crypto dependencies

File: components/filecoin/rs/src/message.rs
Comment (fmarier) [reply components/filecoin/rs/src/message.rs:59]: Is there a reason not to catch & handle properly any errors in converting the string to BigInt?

Comment (fmarier) [reply components/filecoin/rs/src/message.rs:59]: @thypon Not a big deal, but I'm curious why we have two duplicate warnings here...

Comment (thypon) [reply components/filecoin/rs/src/message.rs:59]: If we don't and somehow some not well formatted input will arrive, the whole browser process crash.
Let's fix this

Comment (thypon) [reply components/filecoin/rs/src/message.rs:59]: For some reason the rule is specified twice.
I'll fix this at the end of this month

```
$ ag trailofbits.rs.panic-in-function-returning-result.panic-in-function-returning-result
assets/semgrep_rules/generated/nonfree/audit.yaml
[1] 31464:7:- id: trailofbits.rs.panic-in-function-returning-result.panic-in-function-returning-result
[2] 31483:35:    source: https://semgrep.dev/r/trailofbits.rs.panic-in-function-returning-result.panic-in-function-returning-result
[3] 31489:54:        url: https://semgrep.dev/playground/r/gETRRJ/trailofbits.rs.panic-in-function-returning-result.panic-in-function-returning-result

assets/semgrep_rules/generated/oss/audit.yaml
[4] 33138:7:- id: trailofbits.rs.panic-in-function-returning-result.panic-in-function-returning-result
[5] 33157:35:    source: https://semgrep.dev/r/trailofbits.rs.panic-in-function-returning-result.panic-in-function-returning-result
[6] 33163:54:        url: https://semgrep.dev/playground/r/gETRRJ/trailofbits.rs.panic-in-function-returning-result.panic-in-function-returning-result
```

File: components/filecoin/rs/Cargo.toml
Comment (rillian) [inline components/filecoin/rs/Cargo.toml:19]: Seems like `anyhow` and `bitflags` aren't actually used?

File: components/filecoin/rs/src/signature.rs
Comment (rillian) [inline components/filecoin/rs/src/signature.rs:12]: You can remove this import now that `to_vec` is available.

File: components/filecoin/rs/src/signature.rs
Comment (rillian) [inline components/filecoin/rs/src/signature.rs:16]: New code uses full paths to this type, so you can drop this import as well.

File: components/filecoin/rs/src/signature.rs
Comment (rillian) [inline components/filecoin/rs/src/signature.rs:17]: `SignatureType` doesn't seem to be used. Please remove it so the compiler doesn't warn.

File: components/filecoin/rs/src/signature.rs
Comment (rillian) [inline components/filecoin/rs/src/signature.rs:83]: Should the `64` here be `SIGNATURE_SIZE`?

File: components/filecoin/rs/src/signature.rs
Comment (rillian) [inline components/filecoin/rs/src/signature.rs:132]: Rust doesn't require parentheses around the conditional.

File: components/filecoin/rs/src/signature.rs
Comment (rillian) [inline components/filecoin/rs/src/signature.rs:147]: While we're here, maybe comment that cxx doesn't support Result types, so return an empty string to signal failure.

File: build/rust/Cargo.lock
Comment (bridiver) [inline build/rust/Cargo.lock:349]: what is this needed for? We should not be building cpp files from cargo so this is a banned feature

File: build/rust/Cargo.lock
Comment (bridiver) [inline build/rust/Cargo.lock:1528]: Why is multihash being bumped here? Please remove any Cargo.lock changes that are not strictly necessary for this.

File: build/rust/Cargo.lock
Comment (bridiver) [inline build/rust/Cargo.lock:437]: There is already a version of `constant_time_eq` in Cargo.lock. Is this version necessary, and if so, can we update the other dependencies so they all use the same version?

File: build/rust/Cargo.lock
Comment (bridiver) [inline build/rust/Cargo.lock:707]: is this version bump necessary? Please don't make any Cargo.lock changes that aren't absolutely necessary

File: build/rust/Cargo.lock
Comment (bridiver) [inline build/rust/Cargo.lock:2355]: this bump also doesn't look related so it should be rolled back

File: build/rust/Cargo.lock
Comment (bridiver) [reply build/rust/Cargo.lock:1528]: it looks like rolling this back would eliminate most of the extraneous changes to Cargo.lock here

Comment (rillian) [reply build/rust/Cargo.lock:1528]: multihash 0.16.3 is required by fvm_shared. The last release compatible with our current 0.16.2 is fvm_shared@v0.9.0, released in August.

@cypt4 Do you know if the ripemd hash is necessary to support for the new address type?

File: components/brave_wallet/common/common_util.cc
Comment (supermassive) [inline components/brave_wallet/common/common_util.cc:33]: `value |= 0` What's the purpose of this line?

File: components/brave_wallet/common/common_util.cc
Comment (supermassive) [inline components/brave_wallet/common/common_util.cc:32]: Needs comment with the spec link.
Also needs unit tests based on some public test vectors.

File: components/brave_wallet/common/fil_address.cc
Comment (supermassive) [inline components/brave_wallet/common/fil_address.cc:136]: Can we just compare two bytes instead of conversion to int?

File: components/brave_wallet/common/fil_address.cc
Comment (supermassive) [inline components/brave_wallet/common/fil_address.cc:261]: Can we assume and support only agent_id == 10? Or this is supposed to be extended in nearest future?
If agent_id == 10 always true then do we even need Leb128Encode?

File: components/brave_wallet/browser/fil_requests.cc
Comment (supermassive) [inline components/brave_wallet/browser/fil_requests.cc:122]: Can we assume there could be no other appearances of `{signed_tx}` in json?

File: build/rust/Cargo.lock
Comment (kdenhartog) [reply build/rust/Cargo.lock:437]: I believe if we bump `blake2b_simd` to v1.0.1 then we'd drop the old version of `constant_time_eq` and utilize only this newer version. However, from the looks of it bumping to this version now brings in some unsafe rust because they're trying to utilize assembly code to prevent the compiler from optimizing the code in a way that could break the constant time guarantees it looks like.

I'd prefer to stick with `0.1.5` until I've had more time to dig into what [this code](https://github.com/cesarb/constant_time_eq/compare/0.1.5...0.2.5#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R3-R49) is doing and be sure this is safe.

File: build/rust/Cargo.lock
Comment (kdenhartog) [reply build/rust/Cargo.lock:1528]: it looks like [blake2b](https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0048.md#the-f4-address-class) is used

File: build/rust/Cargo.lock
Comment (cypt4) [reply build/rust/Cargo.lock:2355]: This is auto-bumped since fvm_shared is updated to 3.3.0
https://crates.io/crates/fvm_shared/3.3.0/dependencies
https://crates.io/crates/fvm_ipld_encoding/0.3.3/dependencies

File: components/brave_wallet/common/common_util.h
Comment (supermassive) [inline components/brave_wallet/common/common_util.h:12]: -

File: components/brave_wallet/common/fil_address.cc
Comment (supermassive) [inline components/brave_wallet/common/fil_address.cc:131]: Does it actually needs ToUpper? https://source.chromium.org/chromium/chromium/src/+/main:components/base32/base32.cc;l=23-28?q=ReverseMapping&ss=chromium

File: components/brave_wallet/common/fil_address.cc
Comment (supermassive) [inline components/brave_wallet/common/fil_address.cc:251]: Just `network_ + "410f" + encoded_output` - function comment already explains what it is

**Review by supermassive — APPROVED** [review]
wallet core LGTM with some minor tweaks suggested

**Review by supermassive — APPROVED** [review]
brave-core LGTM

File: components/brave_wallet/common/fil_address.cc
Comment (cypt4) [reply components/brave_wallet/common/fil_address.cc]: Ok, hardcoded 10

File: components/brave_wallet/common/common_util.cc
Comment (cypt4) [reply components/brave_wallet/common/common_util.cc]: Removed

File: components/brave_wallet/common/common_util.cc
Comment (cypt4) [reply components/brave_wallet/common/common_util.cc]: Removed

File: components/brave_wallet/common/fil_address.cc
Comment (cypt4) [reply components/brave_wallet/common/fil_address.cc:131]: Seems letter case is taken into account

File: components/filecoin/rs/Cargo.toml
Comment (fmarier) [inline components/filecoin/rs/Cargo.toml:12]: If we're adding a dependency on this module, it needs to be moved to https://github.com/brave/ so that it's in scope for our bug bounty program. You can review [this checklist](https://github.com/brave/internal/wiki/GitHub-Security-Checklist) and then ask it-support to move it.

cc @kdenhartog who may want to take a look at it too

**Review by fmarier — CHANGES_REQUESTED** [review]

File: components/filecoin/rs/Cargo.toml
Comment (kdenhartog) [reply components/filecoin/rs/Cargo.toml:12]: All of the dependencies that are used in this branch have been audited previously in https://github.com/brave/security/issues/840 and are approved in https://github.com/brave/handbook/blob/master/development/crypto-dependencies.md.

As long as we follow the guidance from @fmarier above this approach LGTM

**Review by muliswilliam — APPROVED** [review]

File: components/brave_wallet/browser/fil_requests.cc
Comment (cypt4) [reply components/brave_wallet/browser/fil_requests.cc:120]: This is not related to the rust signing code. Problem is that signed transaction which is passed as a string everywhere has fields that have uint64 type.
So we inplace this string without converting it to base::Value since this could break some fields(convert them to double).
Similar approach is used in FilTransaction::GetSignedTransaction.
Alternatively we can switch to using base::Value with string uint64 fields when passing signedTx verywhere, but this will require notable refactoring.

File: components/filecoin/rs/Cargo.toml
Comment (kdenhartog) [inline components/filecoin/rs/Cargo.toml:12]:
```suggestion
fvm_shared = { git = "https://github.com/brave/ref-fvm", branch = "3.2.0_downgraded_deps", default-features = false }
```

File: components/filecoin/rs/src/signature.rs
Comment (kdenhartog) [inline components/filecoin/rs/src/signature.rs:1]: I've reviewed the changes in this file for cryptography usage and have no concerns with what is being done. This is using libraries we've already merged into place and the usage does not appear to introduce any cryptographic vulnerabilities.

**Review by kdenhartog — COMMENTED** [review]
Just need to update to point at the new brave repo and this one should be G2G for security review.

cc @stoletheminerals since he'll be taking over the rest of the security review for this one

File: components/brave_wallet/browser/fil_requests.cc
Comment (stoletheminerals) [reply components/brave_wallet/browser/fil_requests.cc:122]: Same question. Could there be a key with an arbitrary value? And if so, are we sure that `{signed_tx}` that we want to replace will alway be the first in the JSON body?

Comment (cypt4) [reply components/brave_wallet/browser/fil_requests.cc:122]: Fields are listed in the method, so there shouldn't be such field

File: components/brave_wallet/browser/fil_requests.cc
Comment (bridiver) [reply components/brave_wallet/browser/fil_requests.cc:120]: I think you're misunderstanding what I'm saying. I don't think this is a good way to convert fields to strings and afaik we already have a way to do this conversion by running it through the rust parser and telling it the field(s) we want to convert

File: components/brave_wallet/browser/fil_requests.cc
Comment (bridiver) [reply components/brave_wallet/browser/fil_requests.cc:122]: this is part of what I'm getting at here https://github.com/brave/brave-core/pull/18511/files#r1210199905
this seems very hacky and fragile and I'm fairly certain we already have a better way to handle this

File: components/brave_wallet/browser/fil_requests.cc
Comment (cypt4) [reply components/brave_wallet/browser/fil_requests.cc:120]: Do you mean something like json::convert_string_value_to_uint64 but to dict instead? There is no such method ATM.

**Review by kdenhartog — APPROVED** [review]
Approving crypto dependencies and signing considerations here. I didn't spot any other concerns either @stoletheminerals. I'll let you close this out in case you're looking at other aspects of this still.

**Review by stoletheminerals — APPROVED** [review]
Lgtm from the security POV

**Review by fmarier — APPROVED** [review]
I didn't review the whole PR but the `fvm_shared` dependency now using a production-ready repo :+1:

File: build/rust/Cargo.lock
Comment (bridiver) [reply build/rust/Cargo.lock]: looking at Cargo.lock for multihash, the only change it should be triggering is adding ripemd and hex-literal. If we can reduce it down to just the 0.16.2 -> 0.16.3 bump and those two additional deps (nothing else should need to be bumped) then it's fine

File: build/rust/Cargo.lock
Comment (bridiver) [reply build/rust/Cargo.lock]: @kdenhartog the issue is not using `constant_time_eq`, it's that we're not using the version we already have in Cargo.lock. We need to avoid having multiple copies of the same dep whenever possible, especially when it brings in transitive deps that have different versions. We need to make the minimum changes to Cargo.lock to satisfy the requirements and not the maximum which is what cargo does by default.

File: build/rust/Cargo.lock
Comment (bridiver) [reply build/rust/Cargo.lock]: Yes, and we should undo that as we have done in other PRs that made changes to Cargo.lock. cargo updates everything to the latest compatible version by default, but we don't want that. We want to update only what is absolutely necessary so you need to leave out changes that we don't actually need and only include the ones that we do need. After cargo runs you need to revert the changes we don't actually need.

Comment (cypt4) [reply build/rust/Cargo.lock]: Wouldn't cargo overwrite them on every launch?

**Review by thypon — APPROVED** [review]

File: build/rust/Cargo.lock
Comment (kdenhartog) [inline build/rust/Cargo.lock:1]: the updated cargo.lock file looks good to me

**Review by kdenhartog — APPROVED** [review]
Thanks for taking the time to try a few different approaches here. This updated cargo.lock approach LGTM

Can we make sure to get the IT team to delete the fork we created as well

File: build/rust/Cargo.lock
Comment (rillian) [reply build/rust/Cargo.lock:2316]: Yes, manual edits of Cargo.lock will have to be redone every time. fvm_ipld_encoding [requires](https://github.com/filecoin-project/ref-fvm/blob/fvm_ipld_encoding%40v0.3.3/ipld/encoding/Cargo.toml#L12) serde_ipld_dagcbor v0.2.2 so cargo will try to reset the back reference.

@bridiver can you confirm that's what you're asking for?

Comment (rillian) [reply build/rust/Cargo.lock:2316]: @kdenhartog how concerned are you about [this issue](https://github.com/ipld/serde_ipld_dagcbor/pull/2)? The functional difference between our 0.2.0 and the 0.2.2 version fvm specifies is that our current version doesn't enforce the identity [multibase prefix](https://github.com/multiformats/multibase#multibase-table) for the cid, allowing base64, etc. encodings to be interpreted as binary.

Comment (rillian) [reply build/rust/Cargo.lock:2316]: @bridiver said on slack that bumps like this were ok, so afaict we've agreed this change is acceptable.

**Review by rillian — APPROVED** [review]
Issues I'm aware of are resolved, so approving for rust-deps. However @bridiver requests this not be merged yet. Please get final approval from him.

File: build/rust/Cargo.lock
Comment (kdenhartog) [reply build/rust/Cargo.lock:2316]: I'd prefer bumping this personally. This seems functionally useful for us.

File: components/filecoin/rs/Cargo.toml
Comment (bridiver) [inline components/filecoin/rs/Cargo.toml:18]: if a more specific minimum version is not required, please default to the epoch `blake2b_simd = "1"`

File: components/filecoin/rs/BUILD.gn
Comment (bridiver) [inline components/filecoin/rs/BUILD.gn:25]: this should be `v3`

File: third_party/rust/fvm_shared/v3_2/BUILD.gn
Comment (bridiver) [inline third_party/rust/fvm_shared/v3_2/BUILD.gn:9]: `epoch = "3"`

**Review by bridiver — APPROVED** [review]

**Comment by srirambv** [comment]
Verification passed on 

Brave | 1.58.77 Chromium: 116.0.5845.51 (Official Build) nightly (64-bit)
-- | --
Revision | 74bdc1d630d841ba7019c2f60c8884b68b1bdb46
OS | Linux

- Verified steps from PR 
- Verified converting ETH to f40 address and doing a send transactions succeeds 
- Verified correct block explorer correctly is opened 
- Verified account has the transaction in the history

https://github.com/brave/brave-core/assets/17010094/c0fe342f-2d4c-4013-9e44-fc8f6eedbf8c

**Pull request merged**
